Add stableHLO->ttir translation of tt.mark custom call

ajakovljevicTT opened the pull request:
As per https://github.com/tenstorrent/tt-mlir/issues/4248, to mark specific inputs as weights for the compiler/tt-metal, we need to pass this information through to the StableHLO mlir code. The idea is for the tt-xla frontend to do that via a custom tt.mark function, which will be annotated with the tt.role attribute, currently only set to "weight". 

That means the generated StableHLO code will look like this:
```
module @jit__lambda_ attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32, tt.mark_function_defined = true} {
  func.func private @tt.mark(%arg0: tensor<4xf32>) -> tensor<4xf32> {
    return %arg0 : tensor<4xf32>
  }
  func.func public @main(%arg0: tensor<4xf32>, %arg1: tensor<3x4xf32>, %arg2: tensor<1x3xf32>) -> (tensor<1x4xf32> {jax.result_info = "result"}) {
    %0 = stablehlo.custom_call @tt.mark(%arg0) {tt.role = "weight"} : (tensor<4xf32>) -> tensor<4xf32>
    %1 = stablehlo.custom_call @tt.mark(%arg1) {tt.role = "weight"} : (tensor<3x4xf32>) -> tensor<3x4xf32>
    %2 = stablehlo.dot_general %arg2, %1, contracting_dims = [1] x [0] : (tensor<1x3xf32>, tensor<3x4xf32>) -> tensor<1x4xf32>
    %3 = stablehlo.reshape %0 : (tensor<4xf32>) -> tensor<1x4xf32>
    %4 = stablehlo.add %2, %3 : tensor<1x4xf32>
    return %4 : tensor<1x4xf32>
  }
}
```

And the resulting ttir should annotate the inputs as needed, giving this result:
```
module @jit__lambda_ attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32, tt.mark_function_defined = true} {
  func.func public @main(%arg0: tensor<4xf32> {tt.role = "weight"}, %arg1: tensor<3x4xf32> {tt.role = "weight"}, %arg2: tensor<1x3xf32>) -> (tensor<1x4xf32> {jax.result_info = "result"}) {
    %0 = "ttir.dot_general"(%arg2, %arg1) <{batch_dims_lhs = array<i64>, batch_dims_rhs = array<i64>, contract_dims_lhs = array<i64: 1>, contract_dims_rhs = array<i64: 0>}> : (tensor<1x3xf32>, tensor<3x4xf32>) -> tensor<1x4xf32>
    %1 = ttir.empty() : tensor<1x4xf32>
    %2 = "ttir.reshape"(%arg0, %1) <{shape = [1 : i32, 4 : i32]}> : (tensor<4xf32>, tensor<1x4xf32>) -> tensor<1x4xf32>
    %3 = ttir.empty() : tensor<1x4xf32>
    %4 = "ttir.add"(%0, %2, %3) : (tensor<1x4xf32>, tensor<1x4xf32>, tensor<1x4xf32>) -> tensor<1x4xf32>
    return %4 : tensor<1x4xf32>
  }
}
```

wooseokTT commented:
@ajakovljevicTT I believe this approach is too aggressive and breaking compatibility of stablehlo. Marking tensor with some attribute may not harm others, so can we do it that way instead of creating custom_call ops?

tapspatel commented:
How are the tt.mark calls inserted? Ie does tt-xla analysis the arguments and insert them? 

There is a `mlir::tt::ttcore::createTTPopulateArgumentTypes(options.argumentTypeMap));` pass which defines types for arguments

see example `test/ttmlir/Dialect/TTIR/populate_argument_types/positive/populate_with_no_existing_attrs.mlir`
```
// RUN: ttmlir-opt --tt-populate-argument-types="argument-types=forward=input,parameter,parameter,constant" -o %t %s
// RUN: FileCheck %s --input-file=%t
module attributes {} {
  // CHECK: ttcore.argument_type = #ttcore.argument_type<input>
  // CHECK: ttcore.argument_type = #ttcore.argument_type<parameter>
  // CHECK: ttcore.argument_type = #ttcore.argument_type<parameter>
  // CHECK: ttcore.argument_type = #ttcore.argument_type<constant>
  func.func @forward(%arg0: tensor<1x32x32x64xbf16>, %arg1: tensor<64x64x3x3xbf16>, %arg2: tensor<1x1x1x64xbf16>, %arg3: tensor<1x32x32x64xbf16>) -> (tensor<1x32x32x64xbf16>, tensor<64x64x3x3xbf16>, tensor<1x1x1x64xbf16>, tensor<1x32x32x64xbf16>) {
    return %arg0, %arg1, %arg2, %arg3 : tensor<1x32x32x64xbf16>, tensor<64x64x3x3xbf16>, tensor<1x1x1x64xbf16>, tensor<1x32x32x64xbf16>
  }
}
```

If we have the argument mapping available in tt-xla, can we do something of how tt-torch does it today?
https://github.com/tenstorrent/tt-torch/blob/79a4f8de9aedd6bea9b759e530b8fe0374440e4c/tt_torch/csrc/tt-mlir-interface.cpp#L113

The stablehlo-pipeline supports this as a parameter: include/ttmlir/Dialect/StableHLO/Pipelines/StableHLOPipelines.h

ajakovljevicTT commented:
> How are the tt.mark calls inserted? Ie does tt-xla analysis the arguments and insert them?
> 
> There is a `mlir::tt::ttcore::createTTPopulateArgumentTypes(options.argumentTypeMap));` pass which defines types for arguments
> 
> see example `test/ttmlir/Dialect/TTIR/populate_argument_types/positive/populate_with_no_existing_attrs.mlir`
> 
> ```
> // RUN: ttmlir-opt --tt-populate-argument-types="argument-types=forward=input,parameter,parameter,constant" -o %t %s
> // RUN: FileCheck %s --input-file=%t
> module attributes {} {
>   // CHECK: ttcore.argument_type = #ttcore.argument_type<input>
>   // CHECK: ttcore.argument_type = #ttcore.argument_type<parameter>
>   // CHECK: ttcore.argument_type = #ttcore.argument_type<parameter>
>   // CHECK: ttcore.argument_type = #ttcore.argument_type<constant>
>   func.func @forward(%arg0: tensor<1x32x32x64xbf16>, %arg1: tensor<64x64x3x3xbf16>, %arg2: tensor<1x1x1x64xbf16>, %arg3: tensor<1x32x32x64xbf16>) -> (tensor<1x32x32x64xbf16>, tensor<64x64x3x3xbf16>, tensor<1x1x1x64xbf16>, tensor<1x32x32x64xbf16>) {
>     return %arg0, %arg1, %arg2, %arg3 : tensor<1x32x32x64xbf16>, tensor<64x64x3x3xbf16>, tensor<1x1x1x64xbf16>, tensor<1x32x32x64xbf16>
>   }
> }
> ```
> 
> If we have the argument mapping available in tt-xla, can we do something of how tt-torch does it today? https://github.com/tenstorrent/tt-torch/blob/79a4f8de9aedd6bea9b759e530b8fe0374440e4c/tt_torch/csrc/tt-mlir-interface.cpp#L113
> 
> The stablehlo-pipeline supports this as a parameter: include/ttmlir/Dialect/StableHLO/Pipelines/StableHLOPipelines.h

@tapspatel THe analysis and insertion is done directly from jax, not tt-xla, and this is the way to pass that information to tt-xla (PJRT) itself. As the next step, we can analyse the MLIR graph gotten from this and fill the ArgumentTypes, but this is just for dealing with the StableHLO code that jax produced.

ajakovljevicTT commented:
> @ajakovljevicTT I believe this approach is too aggressive and breaking compatibility of stablehlo. Marking tensor with some attribute may not harm others, so can we do it that way instead of creating custom_call ops?

@wooseokTT you are correct, thanks for this! I have changed the code to use general "func.call" ops, instead of the "stablehlo.custom_call" ones (and changed the passes appropriatelly), so now the change should not be breaking for stablehlo.

ajakovljevicTT commented:
> > Update: After talking with @wooseokTT and @mrakitaTT offline, we have decided to move this logic to tt-xla, since this will probably be used only there. Hence, closing the PR.
> 
> Can you give us some insights on how this would be implemented in tt-xla? I am just curious to know. :)

The pipeline is more or less the same, it just will live in tt-xla and will be called there, before giving the module to StableHLOtoTTIRPipeline from tt-mlir.
